tasks/Task.cpp

```diff
@@ -28,27 +27,289 @@ bool Task::configureHook()
+        cog[i].set_ij(aux[1], aux[0]);
+    }
+
+    // Compute the coordinates of the 2D reference point from the
+    // intersection of the diagonals of the square
+    double a1 = (cog[0].get_i()- cog[2].get_i()) /
+        (cog[0].get_j()- cog[2].get_j());
+    double a2 = (cog[1].get_i()- cog[3].get_i()) /
+        (cog[1].get_j()- cog[3].get_j());
+    double b1 = cog[0].get_i() - a1*cog[0].get_j();
+    double b2 = cog[1].get_i() - a2*cog[1].get_j();
+
+    refP.set_j( (b1-b2) / (a2-a1) );
+    refP.set_i( a1*refP.get_j() + b1 );
+    double x,y;
+    vpPixelMeterConversion::convertPoint(cam_cal, refP, x, y);
+
+    // Update the 2D ref point visual feature
+    double Z;
+    Z = P.get_Z();
+    p.set_xyZ(x, y, Z);
+    depth.buildFrom(P.get_x(), P.get_y(), Z, log(Z/Zd));
+
+    //////////////////////////////
+    // Update Feature 3 - t, tu //
+    //////////////////////////////
+    cdMc = cdMo*cMo.inverse();
+    t.buildFrom(cdMc);
+    tu.buildFrom(cdMc) ;
+
+}
+
+void Task::updateDesiredPose()
+{
+    // Initialise a desired pose to compute s*, the desired 2D point features
+    base::LinearAngular6DCommand setpoint;
+    _cmd_in.read(setpoint);
+
+    vpTranslationVector cdto(setpoint.linear[0], setpoint.linear[1], setpoint.linear[2]);
+    vpRxyzVector cdro(setpoint.angular[0], setpoint.angular[1], setpoint.angular[2]);
+
+    //Build the rotation and homogeneus matrix according to desired pose
+    vpRotationMatrix cdRo(cdro); // Build the rotation matrix
+    cdMo.buildFrom(cdto, cdRo); // Build the homogeneous matrix
+
+    P.track(cdMo);
+    vpFeatureBuilder::create(pd, P);
+    Zd = P.get_Z();
+}
+
+void Task::setGain()
+{
+    if (_use_adaptive_gain.get())
+    {
+        vpAdaptiveGain  lambda;
+        lambda.initStandard(_adaptive_gains.get().gain_at_zero,
+                _adaptive_gains.get().gain_at_infinity,
+                _adaptive_gains.get().slope_at_zero);
+
+        task.setLambda(lambda) ;
+    }
+    else
+    {
+        task.setLambda(_gain.get());
+    }
+}
+
+void Task::writeVelocities(vpColVector v)
+{
+    base::LinearAngular6DCommand vel;
+    vel.linear[0] = v[0];
+    vel.linear[1] = v[1];
+    vel.linear[2] = v[2];
+    vel.angular[0] = v[3];
+    vel.angular[1] = v[4];
+    vel.angular[2] = v[5];
+    vel.time = base::Time::now();
+
+    _cmd_out.write(vel);
+}
+
+
+void Task::rotX(vpRotationMatrix &Rin, vpRotationMatrix &Rout, double angle_degree)
+{
+    vpRotationMatrix rot(vpMath::rad(angle_degree), 0, 0);
+    Rout = Rin * rot;
+}
+
+void Task::rotY(vpRotationMatrix &Rin, vpRotationMatrix &Rout, double angle_degree)
+{
+    vpRotationMatrix rot(0, vpMath::rad(angle_degree), 0);
+    Rout = Rin * rot;
+}
+
+void Task::rotZ(vpRotationMatrix &Rin, vpRotationMatrix &Rout, double angle_degree)
+{
+    vpRotationMatrix rot(0, 0, vpMath::rad(angle_degree));
+    Rout = Rin * rot;
+}
+
+void Task::poseDisplay(vpHomogeneousMatrix& M)
```

Comment:
*No* *Way*. If you want to output debug data, create a debug structure and a debug port. Which among other things ensures that you can display it properly using e.g. rock-display